azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -1349,16 +1349,24 @@ def on_complete(op, error):
                 this.state = ReconnectState.DISCONNECTED
 
             # Allow the next waiting op to proceed (if any)
-            this._run_next_waiting_op()
+            return this._run_all_waiting_ops()
 
         op.add_callback(on_complete)
 
     @pipeline_thread.runs_on_pipeline_thread
-    def _run_next_waiting_op(self):
+    def _run_all_waiting_ops(self):
+
         if not self.waiting_ops.empty():
-            next_op = self.waiting_ops.get_nowait()
-            logger.debug("{}: Resolving next waiting op: {}".format(self.name, next_op.name))
-            self.run_op(next_op)
+            queuecopy = self.waiting_ops
+            self.waiting_ops = queue.Queue()
+
+            while not queuecopy.empty():
+                next_op = queuecopy.get_nowait()
+                if not next_op.completed:
```

Comment:
Is this condition necessary? I don't think it's possible to be completed and in the queue

Reply:
"don't think it can happen" or "can guarantee that it won't happen"?  Calling `run_op` on a completed op can have unpredictable results, and I'd rather have an extra check here.